src/server/worker/pipeline/service/service.go

```diff
@@ -0,0 +1,123 @@
+package service
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsconsts"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsutil"
+	"github.com/pachyderm/pachyderm/src/server/worker/common"
+	"github.com/pachyderm/pachyderm/src/server/worker/datum"
+	"github.com/pachyderm/pachyderm/src/server/worker/driver"
+	"github.com/pachyderm/pachyderm/src/server/worker/logs"
+	"github.com/pachyderm/pachyderm/src/server/worker/pipeline"
+)
+
+type serviceItem struct {
+	serviceCtx context.Context
+	commitInfo *pfs.CommitInfo
+}
+
+// Runs the given callback with the latest commit for the pipeline.  The given
+// context will be canceled if a newer commit is ready.
+func forLatestCommit(
+	pachClient *client.APIClient,
+	pipelineInfo *pps.PipelineInfo,
+	cb func(context.Context, *pfs.CommitInfo) error,
+) error {
+	// These are used to cancel the existing service and wait for it to finish
+	var cancel func()
+	var eg *errgroup.Group
+
+	return pachClient.SubscribeCommitF(
+		pipelineInfo.Pipeline.Name,
+		"",
+		client.NewCommitProvenance(ppsconsts.SpecRepo, pipelineInfo.Pipeline.Name, pipelineInfo.SpecCommit.ID),
+		"",
+		pfs.CommitState_READY,
+		func(ci *pfs.CommitInfo) error {
+			if cancel != nil {
+				cancel()
+				if err := eg.Wait(); err != nil && err != context.Canceled {
+					return err
+				} else if common.IsDone(pachClient.Ctx()) {
+					return pachClient.Ctx().Err()
+				}
+			}
+
+			var ctx context.Context
+			ctx, cancel = context.WithCancel(pachClient.Ctx())
+			eg, ctx = errgroup.WithContext(ctx)
+			eg.Go(func() error { return cb(ctx, ci) })
+
+			return nil
+		},
+	)
+}
+
+// Run will run a service pipeline until the driver is canceled.
+func Run(driver driver.Driver, logger logs.TaggedLogger) error {
```

Comment:
Where is the check for commits that have already been finished? I don't see it in `forLatestCommit` or here.